src/rpc/blockchain.cpp

```diff
@@ -1423,6 +1423,77 @@ UniValue z_gettreestate(const UniValue& params, bool fHelp)
     return res;
 }
 
+UniValue z_getsubtreesbyindex(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 2 || params.size() > 3)
+        throw runtime_error(
+            "z_getsubtreesbyindex \"pool\" start_index ( end_index )\n"
+            "Return information about the given block's tree state.\n"
+            "\nArguments:\n"
+            "1. \"pool\"        (string, required) The pool from which subtrees should be returned. Either \"sapling\" or \"orchard\".\n"
+            "2. start_index   (numeric, required) The index of the first subtree to return.\n"
+            "2. end_index     (numeric, optional) The index of the last subtree to return.\n"
+            "\nResult:\n"
+            "{\n"
+            "  \"pool\" : \"sapling|orchard\", (string) The shielded pool to which the subtrees belong\n"
+            "  \"depth\": n,            (numeric) The depth of the subtrees within the pool's note commitment tree\n"
+            "  \"start_index\": n,      (numeric) The index of the first subtree\n"
+            "  \"subtrees\": [          (array) A sequential list of complete subtrees\n"
+            "    {\n"
+            "      \"root\": \"hash\",    (string) hash of most recent block with more information\n"
+            "      \"end_height\": n,   (numeric) height of the block containing the note that completed this subtree\n"
+            "    }, ...\n"
+            "  ]\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_getsubtreesbyindex", "\"sapling\", 0")
+            + HelpExampleRpc("z_getsubtreesbyindex", "\"orchard\", 3, 7")
+        );
+
+    auto strPool = params[0].get_str();
+    ShieldedType pool;
+    if (strPool == "sapling") {
+        pool = SAPLING;
+    } else if (strPool == "orchard") {
+        pool = ORCHARD;
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Requested pool must be \"sapling\" or \"orchard\"");
+    }
+
+    libzcash::SubtreeIndex startIndex = params[1].get_int();
+    std::optional<libzcash::SubtreeIndex> endIndex = std::nullopt;
+    if (params.size() > 2) {
+        endIndex = params[2].get_int();
+    }
+
+    LOCK(cs_main);
+
+    UniValue subtrees(UniValue::VARR);
+    for (libzcash::SubtreeIndex index = startIndex; ; index++) {
+        if (endIndex.has_value() && endIndex.value() < index) {
+            break;
+        }
+
+        auto subtreeData = pcoinsTip->GetSubtreeData(pool, index);
+        if (!subtreeData.has_value()) {
+            break;
+        }
+
+        UniValue subtree(UniValue::VOBJ);
+        subtree.pushKV("root", HexStr(subtreeData->root));
+        subtree.pushKV("end_height", subtreeData->nHeight);
+        subtrees.push_back(subtree);
+    }
+
+    UniValue res(UniValue::VOBJ);
+    res.pushKV("pool", strPool);
+    res.pushKV("depth", libzcash::TRACKED_SUBTREE_HEIGHT);
```

Comment:
I defaulted to naming this field `depth`, but we should ensure it is consistent with the naming in the `shardtree` crate.

Reply:
`depth` is typically used for depth in the chain; I think that something like `subtree_height` would be more explicit.

Reply:
Actually, `depth` is fine here.

Reply:
How can "height" mean the same thing as "depth"? If we take the tree to have the root at the top (compatible with the usage of "altitude" in [ZIP 221](https://zips.z.cash/zip-0221)), then "height" would count from the leaves while "depth" would count from the root. They both happen to be 16, but that just makes things worse!

Reply:
Either way we could be confusing multiple things. `depth` could be "depth of the subtrees within the larger commitment tree" (which is what's documented in the help message) or "depth of the subtree itself". We just need to make sure whatever we mean is documented.

Reply:
I think what we actually want is the height of the subtrees; the tricky thing is that we use `height` as a suffix that means "block height" in a whole bunch of places.

Reply:
This is why I prefer "altitude". (We had a bunch of discussion about these ambiguities when writing ZIP 221, iirc.)